metadata/utils.go

```diff
@@ -140,6 +140,14 @@ func GetMetadataProperty(props map[string]string, keys ...string) (val string, o
 // DecodeMetadata decodes metadata into a struct
 // This is an extension of mitchellh/mapstructure which also supports decoding durations
 func DecodeMetadata(input interface{}, result interface{}) error {
+	// avoids a common mistake of passing the metadata object, instead of the properties map
+	// if input is not of type map[string]string, then cast to metadata.Base and access the Properties
+	if _, ok := input.(map[string]string); !ok {
```

Comment:
Why not just changing the signature to `input map[string]string`?

Reply:
Now it's for convenience - being able to support the metadata.base class is actually much nicer.

Reply:
One very convenient option (that would involve some refactoring) would be to change `GetComponentMetadata()` to return `any`. The component would simply return whatever metadata struct they use (for example `return redisMetadata{}`), and the caller would be doing all the reflection magic.

(Bonus: move the code that does `DecodeMetadata` into a file compiled only with `//go:build +metadata` so it's not even included in the built `daprd`)...

Reply:
Sorry, not going to do that in this PR.

Reply:
To be clear I was not suggesting any change in this PR. Just thinking out loud, maybe for the future!